hashicorp/consul
# Sync cluster attributes from GNM to Link resource

### Description


When a Link resource is created, we want to sync certain attributes, such as the HCP Portal URL and the Access Mode from GNM to the Link resource.

This change makes it so the Link controller queries GNM for the cluster, then syncs the HCP Portal URL to the Link resource.

### Testing & Reproduction steps



* Added to existing HCP Link Controller unit tests to verify the new behavior

#### Manual testing - Happy path
1. Create an HCP Link using the HTTP API
```
➜  http PUT localhost:8500/api/hcp/v2/link/global data[resourceId]=$RESOURCE_ID data[clientSecret]=$CLIENT_SECRET data[clientId]=$CLIENT_ID X-Consul-Token:$CONSUL_HTTP_TOKEN
HTTP/1.1 200 OK
Content-Length: 622
Content-Type: text/plain; charset=utf-8
Date: Tue, 16 Jan 2024 22:46:44 GMT

{
    "data": {
        "clientId": "WBV222LhQ9qVRxgGV8rkJIidtKzYRbTK",
        "clientSecret": "<redacted>",
        "resourceId": "organization/d95434ef-1f14-4b17-b23e-1eb608bb9fda/project/e263159f-b9ea-468e-8dc5-ee3d23403fe2/hashicorp.consul.global-network-manager.cluster/test-sync-stuff"
    },
    "generation": "01HMA798PHZMPPYJPWC42ZVF1Z",
    "id": {
        "name": "global",
        "tenancy": {
            "peerName": "local"
        },
        "type": {
            "group": "hcp",
            "groupVersion": "v2",
            "kind": "Link"
        },
        "uid": "01HMA798PHZMPPYJPWC36CNNWB"
    },
    "version": "22"
}

```

2. Retrieve the link afterwards and verify the HCPPortalUrl attribute has been set
```


➜   http GET localhost:8500/api/hcp/v2/link/global X-Consul-Token:$CONSUL_HTTP_TOKEN
HTTP/1.1 200 OK
Content-Length: 1354
Content-Type: text/plain; charset=utf-8
Date: Tue, 16 Jan 2024 22:48:04 GMT

{
    "data": {
        "accessLevel": "ACCESS_LEVEL_GLOBAL_READ_WRITE",
        "clientId": "WBV222LhQ9qVRxgGV8rkJIidtKzYRbTK",
        "clientSecret": "<redacted>",
        "hcpClusterUrl": "https://portal.cloud.hashicorp.com/services/consul/clusters/self-managed/test-sync-stuff?project_id=e263159f-b9ea-468e-8dc5-ee3d23403fe2",
        "resourceId": "organization/d95434ef-1f14-4b17-b23e-1eb608bb9fda/project/e263159f-b9ea-468e-8dc5-ee3d23403fe2/hashicorp.consul.global-network-manager.cluster/test-sync-stuff"
    },
    "generation": "01HMA79M7HZ9BFM7GCNA31EA00",
    "id": {
        "name": "global",
        "tenancy": {
            "peerName": "local"
        },
        "type": {
            "group": "hcp",
            "groupVersion": "v2",
            "kind": "Link"
        },
        "uid": "01HMA798PHZMPPYJPWC36CNNWB"
    },
    "status": {
        "consul.io/hcp/link": {
            "conditions": [
                {
                    "message": "Successfully linked to cluster 'organization/d95434ef-1f14-4b17-b23e-1eb608bb9fda/project/e263159f-b9ea-468e-8dc5-ee3d23403fe2/hashicorp.consul.global-network-manager.cluster/test-sync-stuff'",
                    "reason": "SUCCESS",
                    "state": "STATE_TRUE",
                    "type": "linked"
                }
            ],
            "observedGeneration": "01HMA79M7HZ9BFM7GCNA31EA00",
            "updatedAt": "2024-01-16T22:46:57.474767Z"
        }
    },
    "version": "25"
}


```

#### Manual testing - Invalid client secret

1. Created a link using the HTTP API with an invalid client secret:
```
 consul git:(nickcellino/query-gnm-cluster-hcp-client) ✗ http PUT localhost:8500/api/hcp/v2/link/global data[resourceId]=$RESOURCE_ID data[clientSecret]=WRONG data[clientId]=$CLIENT_ID X-Consul-Token:$CONSUL_HTTP_TOKEN
HTTP/1.1 200 OK
Content-Length: 555
Content-Type: text/plain; charset=utf-8
Date: Fri, 12 Jan 2024 21:29:46 GMT

{
    "data": {
        "clientId": "WBV222LhQ9qVRxgGV8rkJIidtKzYRbTK",
        "clientSecret": "WRONG",
        "resourceId": "organization/d95434ef-1f14-4b17-b23e-1eb608bb9fda/project/e263159f-b9ea-468e-8dc5-ee3d23403fe2/hashicorp.consul.global-network-manager.cluster/hello-world"
    },
    "generation": "01HKZS9F201G2CD2J6A7NYAX5M",
    "id": {
        "name": "global",
        "tenancy": {
            "peerName": "local"
        },
        "type": {
            "group": "hcp",
            "groupVersion": "v2",
            "kind": "Link"
        },
        "uid": "01HKZS9F201G2CD2J6A5SSSYMK"
    },
    "version": "22"
}

```

2. Verify that link is returned in a failed state
```
consul git:(nickcellino/query-gnm-cluster-hcp-client) ✗ http GET localhost:8500/api/hcp/v2/link/global X-Consul-Token:$CONSUL_HTTP_TOKEN
HTTP/1.1 200 OK
Content-Length: 966
Content-Type: text/plain; charset=utf-8
Date: Fri, 12 Jan 2024 21:32:56 GMT

{
    "data": {
        "clientId": "WBV222LhQ9qVRxgGV8rkJIidtKzYRbTK",
        "clientSecret": "WRONG",
        "resourceId": "organization/d95434ef-1f14-4b17-b23e-1eb608bb9fda/project/e263159f-b9ea-468e-8dc5-ee3d23403fe2/hashicorp.consul.global-network-manager.cluster/hello-world"
    },
    "generation": "01HKZSDTXVKHR31M4VN1XEJ1JR",
    "id": {
        "name": "global",
        "tenancy": {
            "peerName": "local"
        },
        "type": {
            "group": "hcp",
            "groupVersion": "v2",
            "kind": "Link"
        },
        "uid": "01HKZSDTXVKHR31M4VN1NC77ZX"
    },
    "status": {
        "consul.io/hcp/link": {
            "conditions": [
                {
                    "message": "Failed to link to HCP",
                    "reason": "FAILED",
                    "state": "STATE_FALSE",
                    "type": "linked"
                }
            ],
            "observedGeneration": "01HKZSDTXVKHR31M4VN1XEJ1JR",
            "updatedAt": "2024-01-12T21:32:40.426647Z"
        }
    },
    "version": "23"
}
```


### Links


Jira: https://hashicorp.atlassian.net/browse/CC-7047

### PR Checklist
